### PR TITLE
Split createCustomer and createCard, mega issue fix

### PR DESCRIPTION
**Pull request by delatbabel**
This PR splits the former createCard functionality (which actually created a customer) into separate createCustomer and createCard requests, as well as fixing issues related to using a customer ID as if it was a card ID.

The stripe gateway actually requires a customer to be created before storing a card in the gateway.  This is not unusual -- many other gateways also require this, e.g. Pin.  I have started a topic for discussion on the omnipay google group about this here: https://groups.google.com/forum/#!topic/omnipay/8Ujt6nX1DKk -- the best way to handle this might be to add additional code to Omnipay Common to determine when a gateway requires a customer object to be created whenever a card object is required.  e.g. adding a new function such as gatewayRequiresCustomerBeforeCard(), etc.

The functionality now allows additional card objects to be added under an existing customer object using CreateCard.  UpdateCustomer now has the ability to change and/or replace an existing customer's default card, as well as the ability to change other customer details if required, e.g. description, email.

I have left behind a backwards compatibility shim whereby if CreateCard is called with card details and no customer reference then a new customer is created.  Both the customer reference and the card reference can now be extracted separately from the return response.  The potential backwards compatibility break caused by purchase() on a card reference now also requring a customer reference as well as the card reference is a non-issue because it never worked before anyway (since attempting to provide the customer reference as a card reference, or providing a card reference with no customer reference, isn't valid -- see issue #6 ).

This fixes issue #8 because it changes the way that updateCard works -- it no longer updates a customer but now correctly updates a card, part of which can be updated.

This fixes issue #6 because it now supports both a customer reference and a card reference on a purchase or authorize request.

I have also added a fix for issue #7 by returning the transaction ID for a failed payment.  I believe that this is unique to stripe but may be useful for some use cases, e.g. retrying or logging failed payments, etc.

Additional test cases and documentation were added -- mostly by moving *Card functionality cases to *Customer and adding additional cases for *Card.  Test case mock responses have been validated against the gateway, which has recently changed some output parameters but not in a way that disrupts backwards compatibility with any of the test cases here.

Lastly -- sorry about you having to read all of the above.  I write long documents because I don't have time to write short ones.

TL;DR It was brokeded before, I fixded it.

### CODE REVIEW

**Comment by kayladnls** [comment]
Splitting the create card functionality into two functions is slated for V3 of Omnipay, but isn't the kind of thing that should be implemented differently in each gateway. 

Also, this is a huge amount of changes for a single PR. Ideally the comments/documentation changes would come in a separate PR from the code changes.

**Comment by delatbabel** [comment]
The problem with making the documentation changes a separate PR is that they make no sense without the code changes.  I have already incorporated all of the documentation changes that make sense with the current code in my previous PR including things like commenting "this class does not work", etc.  This batch of documentation changes really only include notes and examples on how the included bunch of code changes work -- and I would be reluctant to have them merged separately to the code changes because in that case they would then be incorrect.

Perhaps this can be put onto a branch waiting a v3 specification?  There are already a couple of gateways that include createCustomer functionality, it would just be a matter of going over them to make sure that the APIs were consistent.

I wasn't aware that there was a v3 roadmap -- is it available publically anywhere?

**Comment by delatbabel** [comment]
I should also point out that the additional functionality for adding customers and cards separately duplicates that which is already in the omnipay-pin gateway code.  Pin is another gateway that doesn't support card creation without a customer (or more exactly, the card that is created can only be used once, to add to a customer) and therefore needs the same fix.  That functionality has been in omnipay-pin since July 2014 so we have a precedent. :)

**Comment by kayladnls** [comment]
That is a good point. Give me a bit for a review and I may be able to get this merged in. It will be a good blueprint for when I update Common and push the functionality to other gateways.

**Comment by kayladnls** [comment]
What I meant about the documentation is that this PR has a lot of documentation that's not central to the feature being implemented. As such, I'm not able to merge the other docs without merging all of it.

**Comment by delatbabel** [comment]
OK I see what you mean, there is some documentation that's just documenting existing things. Sorry, I thought I had PR'ed that already, I must have missed it in a flurry of feathers.

Having said that there are some docblocks that don't look like they are central to this functionality but in reality are.  For example, this one:

```
+ * Either a customerReference or a card is required. If a customerReference
+ * is passed in then the cardReference must be the reference of a card
+ * assigned to the customer. Otherwise, if you do not pass a customer ID,
+ * the card you provide must either be a token, like the ones returned by
+ * Stripe.js, or a dictionary containing a user's credit card details.
+ *
+ * IN OTHER WORDS: You cannot just pass a card reference into this request,
+ * you must also provide a customer reference if you want to use a stored
+ * card.
+ *
```

although in theory you could write something about the previous incarnation of that message, in fact without both a customer ID and a card ID the function actually won't work (the previous incarnation passed just a customer ID which will in fact fail).  I had avoided docblocking that function previously because there was no way it could work as "advertised" (at least by the Omnipay common API).

So I could go and pull this PR apart into separate PRs for existing functionality and modified functionality, however a lot of the docblocks for the existing functions will probably just have to say "this function does not work, do not use it". Which is more or less what I did when docblocking the CreateCardRequest message -- I just referenced the issue which in turn had notes about the problem with the message class. It's also a fair amount of work for me and to be honest my git/github legs are still a little wet and wobbly so I'd be more concerned that I'd mess something else up if I did that.

Also it's Vietnamese new year (where I live at the moment) so I won't get to anything else on this until next week.

**Comment by delatbabel** [comment]
I also have to say I'm probably creating a bundle of work for you even reviewing this stuff, I appreciate of all of the work you do in maintaining these packages in some semblance of order.  If this has to wait until your real life or work allows you to check it over or if you have higher priority things to do then so be it.

**Comment by kayladnls** [comment]
I reviewed it yesterday and it all looked clean and well documented, with the huge added plus of actually working, as opposed to the previous version which didn't. I'll be merging tonight. 

And thanks a ton for all your work on these packages. It's really invaluable to me.

**Comment by delatbabel** [comment]
Hi, Any progress on this? I'm still having to use a VCS pull of my own repo because I need to work with stored cards.

**Comment by kayladnls** [comment]
Sorry about the delay on this @delatbabel, life's gotten in the way. I can rebase and merge over the weekend. If you're able to rebase to get it up to date, I can merge now.

**Comment by delatbabel** [comment]
This commit rebases the change with respect to the current head code.  I also resolved some test issues relating to PR #22

**Comment by delatbabel** [comment]
We seem to have dropped a bit of code coverage, in particular setters and getters for source/destination parameters added in #22 appear not covered at the moment.  Just FYI -- I will try to get to these at some stage, or you can ask qlik if he can do them for his code after merging this.

**Comment by delatbabel** [comment]
Also added the same change as #13 as well as some tests.  I also added test to cover test cases for get/set Source that were missing.

File: src/Message/AbstractRequest.php [hunk 1]
```diff
@@ -154,21 +177,24 @@ public function setSource($value)
...
-        $data['address_line1'] = $this->getCard()->getAddress1();
-        $data['address_line2'] = $this->getCard()->getAddress2();
-        $data['address_city'] = $this->getCard()->getCity();
-        $data['address_zip'] = $this->getCard()->getPostcode();
-        $data['address_state'] = $this->getCard()->getState();
-        $data['address_country'] = $this->getCard()->getCountry();
+        $data['number'] = $card->getNumber();
+        $data['exp_month'] = $card->getExpiryMonth();
+        $data['exp_year'] = $card->getExpiryYear();
+        if ($this->getCard()->getCvv()) {
```
Comment (evgenius) [inline src/Message/AbstractRequest.php]: Why isn't that just `if ($card->getCvv())`?

Comment (delatbabel) [inline src/Message/AbstractRequest.php]: Yes, you're correct, laziness from copying someone else's code.  I will fix this shortly.

**Comment by phred** [comment]
As of the [Feb 18, 2015 update to the Stripe API](https://stripe.com/docs/upgrades#2015-02-18):

> The `cards` and `default_card` attributes are no longer returned on Customers. You should now use `sources` and `default_source` respectively. The `customer.card.*` and `customer.bank_account.*` webhooks are now named `customer.source.*`. If you only have cards attached to customers (as opposed to payment sources of other types) then you can use the new attributes exactly as you would the old ones. If you have payment sources of multiple types, then the sources list contains heterogeneous objects and you can check the object attribute of each source to determine its format. Older API versions return both the new and the old attributes.

This will definitely cause problems if you're running on the latest Stripe API, which I assume is the default for new Stripe accounts.

**Comment by kayladnls** [comment]
Thanks again @delatbabel
